### ISSUE TITLE
`str.json_decode` hangs on nested JSON structure with null values

### ISSUE THREAD

**Issue by nsfinkelstein** [pola-rs/polars#18781]
### Checks

- [X] I have checked that this issue has not already been reported.
- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Reproducible example

```python
import polars as pl
pl.Series([
    '{"a":[{"b":false}]}',
    '{"a":[{"b":null}]}',
    '{"a":null}',
]).str.json_decode()
```

### Issue description

The above and related code hangs interminably.

This code runs correctly on polars version `1.6.0`, with all other versions the same.

### Expected behavior

```
shape: (3,)
Series: '' [struct[1]]
[
        {[{false}]}
        {[{null}]}
        {null}
]
```

### Installed versions

<details>

```
--------Version info---------
Polars:              1.7.1
Index type:          UInt32
Platform:            Linux-5.10.219-208.866.amzn2.x86_64-x86_64-with-glibc2.35
Python:              3.11.9 (main, Sep  8 2024, 07:11:41) [GCC 11.4.0]

----Optional dependencies----
adbc_driver_manager  <not installed>
altair               <not installed>
cloudpickle          3.0.0
connectorx           <not installed>
deltalake            <not installed>
fastexcel            <not installed>
fsspec               2024.9.0
gevent               <not installed>
great_tables         <not installed>
matplotlib           3.9.2
nest_asyncio         1.6.0
numpy                1.26.4
openpyxl             <not installed>
pandas               2.2.2
pyarrow              17.0.0
pydantic             2.8.2
pyiceberg            <not installed>
sqlalchemy           2.0.34
torch                2.3.1+cu121
xlsx2csv             <not installed>
xlsxwriter           <not installed>
```

</details>

**Comment by cmdlineluser** [comment]
This segfaults for me with a debug build.

```
zsh: segmentation fault
```

It seems to have started after this PR:

- https://github.com/pola-rs/polars/pull/18425

**Comment by cmdlineluser** [comment]
This is still reproducible as of 1.8.1

Hope the ping is OK @ritchie46 - but it seems like this one may need attention if it is not on the radar already.

**Comment by ritchie46** [comment]
Ah, I missed it. But.... I coincidentally encountered and fixed it myself! :D

Fixed in #18887
